File: dungeons-26.1.2/src/main/java/fishmod/utils/dungeon/DungeonClass.java
package fishmod.utils.dungeon;

import fishmod.utils.Misc;
import fishmod.utils.config.values.Dungeons;
import fishmod.utils.data.EntityUtil;
import fishmod.utils.events.Events;
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.player.Player;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public enum DungeonClass {
    ARCHER, BERSERK, HEALER, MAGE, TANK;

    private static final Pattern PATTERN = Pattern.compile("^\\[(Archer|Berserk|Healer|Mage|Tank)]");
    private static final Pattern NAME_CLASS_PATTERN = Pattern.compile("^\\[\\d+] (.+) \\((Archer|Berserk|Healer|Mage|Tank) ");
    /** "Your Berserk stats are doubled because you are the only player using this class!" — the most
     *  reliable signal of the LOCAL player's own dungeon class (fires at the start of a run). */
    private static final Pattern STATS_DOUBLED_PATTERN =
            Pattern.compile("Your (Archer|Berserk|Healer|Mage|Tank) stats are doubled because you are the only player using this class!");

    private static final ConcurrentHashMap<String, DungeonClass> nameClassMap = new ConcurrentHashMap<>();
    public static DungeonClass currentClass;

    public static void init() {

        Events.ON_PHASE_CHANGE.register(() -> {
            if (Phase.runJustStarted()) {
                reset();
            }
            return false;
        });

        Events.ON_RUN_END.register(() -> {
            reset();
            return false;
        });

        ClientReceiveMessageEvents.GAME.register((message, overlay) -> {
            String string = message.getString();

            // Authoritative own-class signal — always wins, regardless of run state.
            Matcher doubled = STATS_DOUBLED_PATTERN.matcher(string);
            if (doubled.find()) {
                currentClass = parseClass(doubled.group(1));
                return;
            }

            if (!Phase.runStarted() && currentClass != null) return;

            // Fallback only: the dungeon chat "[Class]" prefix is the SPEAKER's class, not necessarily
            // ours — only use it to seed currentClass when nothing more reliable has set it yet.
            Matcher matcher = PATTERN.matcher(string);
            if (matcher.find() && currentClass == null) {
                currentClass = parseClass(matcher.group(1));
            }
        });

        Events.ON_PLAYER_ENTRY.register(receivedEntry -> {
            if (receivedEntry == null) return false;
            Component text = receivedEntry.displayName();
            if (text == null) return false;
            String string = text.getString();
            Matcher matcher = NAME_CLASS_PATTERN.matcher(string);

            if (matcher.find()) {
                String name = matcher.group(1).replaceAll(" .+", "");
                DungeonClass className = parseClass(matcher.group(2));
                if (className == null) return false;

                if (EntityUtil.isClientPlayer(name)) {
                    currentClass =  className;
                }

                nameClassMap.put(name, className);
                return false;
            }

            return false;
        });

    }

    private static DungeonClass parseClass(String name) {
        try {
            return DungeonClass.valueOf(name.toUpperCase());
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    public static DungeonClass getClass(String playerName) {
        if (playerName == null) return null;
        if (nameClassMap.containsKey(playerName)) {
            return nameClassMap.get(playerName);
        }
        return null;
    }

    public static DungeonClass getClass(Player player) {
        if (player == null) return null;
        return getClass(player.getName().getString());
    }

    public static boolean isTeammate(Player player) {
        if (player == null || EntityUtil.isClientPlayer(player)) return false;
        String name = player.getName().getString();
        return nameClassMap.containsKey(name);
    }

    public static int getColor(DungeonClass dungeonClass) {
        if (dungeonClass == null) return 0xffffffff;

        return switch (dungeonClass) {
            case ARCHER -> Dungeons.archerColor;
            case BERSERK -> Dungeons.berserkColor;
            case HEALER -> Dungeons.healerColor;
            case MAGE -> Dungeons.mageColor;
            case TANK -> Dungeons.tankColor;
        };
    }

    public static String getChar(DungeonClass dungeonClass) {
        if (dungeonClass == null) return "?";

        return switch (dungeonClass) {
            case ARCHER -> "A";
            case BERSERK -> "B";
            case HEALER -> "H";
            case MAGE -> "M";
            case TANK -> "T";
        };
    }

    public static int getColor(String name) {
        return getColor(getClass(name));
    }

    public static boolean isClass(DungeonClass dungeonClass) {
        return currentClass == dungeonClass;
    }

    public static boolean isArchTeam() {
        return currentClass == DungeonClass.ARCHER || currentClass == DungeonClass.TANK;
    }

    public static boolean isBersTeam() {
        return currentClass == DungeonClass.BERSERK || currentClass == DungeonClass.MAGE;
    }

    private static void reset() {
        currentClass = null;
        nameClassMap.clear();
    }

    public static void printClasses() {
        Misc.addChatMessage(Component.literal("Classes"));
        nameClassMap.forEach((name, clazz) -> {
            Misc.addChatMessage(Component.literal("Name: " + name + "Class: " + (clazz != null? clazz.name(): null)));
        });
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/utils/dungeon/DungeonClass.kt
package fishmod.utils.dungeon

import fishmod.utils.Misc
import fishmod.utils.config.values.Dungeons
import fishmod.utils.data.EntityUtil
import fishmod.utils.events.Events
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents
import net.minecraft.network.chat.Component
import net.minecraft.world.entity.player.Player
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Pattern

enum class DungeonClass {
    ARCHER, BERSERK, HEALER, MAGE, TANK;

    companion object {
        private val PATTERN: Pattern = Pattern.compile("^\\[(Archer|Berserk|Healer|Mage|Tank)]")
        private val NAME_CLASS_PATTERN: Pattern = Pattern.compile("^\\[\\d+] (.+) \\((Archer|Berserk|Healer|Mage|Tank) ")

        /** Most reliable signal of the local player's own class — fires at the start of a run. */
        private val STATS_DOUBLED_PATTERN: Pattern =
            Pattern.compile("Your (Archer|Berserk|Healer|Mage|Tank) stats are doubled because you are the only player using this class!")

        private val SELECTED_PATTERN: Pattern =
            Pattern.compile("You have selected the (Archer|Berserk|Healer|Mage|Tank) Dungeon Class!")

        private val nameClassMap = ConcurrentHashMap<String, DungeonClass>()

        @JvmField
        var currentClass: DungeonClass? = null

        @JvmStatic
        fun init() {

            // Clear the class map when we actually load a new instance / the run ends — NOT on
            // every phase change during clear (runJustStarted() stays true the whole clear phase,
            // so that wiped everyone's class mid-run and only some refilled from the tab list).
            Events.ON_WORLD_CHANGE.register { reset(); false }
            Events.ON_RUN_END.register { reset(); false }

            ClientReceiveMessageEvents.GAME.register { message, _ ->
                val string = message.string

                // Authoritative own-class signals — always win, regardless of run state.
                val selected = SELECTED_PATTERN.matcher(string)
                if (selected.find()) {
                    currentClass = parseClass(selected.group(1))
                    return@register
                }
                val doubled = STATS_DOUBLED_PATTERN.matcher(string)
                if (doubled.find()) {
                    currentClass = parseClass(doubled.group(1))
                    return@register
                }

                if (!Phase.runStarted() && currentClass != null) return@register

                // the "[Class]" chat prefix is the SPEAKER's class, not necessarily ours — only a fallback seed
                val matcher = PATTERN.matcher(string)
                if (matcher.find() && currentClass == null) {
                    currentClass = parseClass(matcher.group(1))
                }
            }

            Events.ON_PLAYER_ENTRY.register { receivedEntry ->
                if (receivedEntry == null) return@register false
                val text = receivedEntry.displayName() ?: return@register false
                val string = text.string
                val matcher = NAME_CLASS_PATTERN.matcher(string)

                if (matcher.find()) {
                    // group(1) is everything between "[lvl] " and " (Class"; it may carry rank tags
                    // ("[YOUTUBE] Future77"), so the IGN is the last whitespace-separated token.
                    val name = matcher.group(1).trim().substringAfterLast(' ')
                    val className = parseClass(matcher.group(2))
                    if (className == null) return@register false

                    if (EntityUtil.isClientPlayer(name)) {
                        currentClass = className
                    }

                    nameClassMap[name] = className
                    return@register false
                }

                false
            }
        }

        private fun parseClass(name: String): DungeonClass? {
            return try {
                valueOf(name.uppercase())
            } catch (e: IllegalArgumentException) {
                null
            }
        }

        @JvmStatic
        fun getClass(playerName: String?): DungeonClass? {
            if (playerName == null) return null
            return nameClassMap[playerName]
        }

        @JvmStatic
        fun getClass(player: Player?): DungeonClass? {
            if (player == null) return null
            return getClass(player.name.string)
        }

        @JvmStatic
        fun getAll(): Map<String, DungeonClass> = Collections.unmodifiableMap(nameClassMap)

        @JvmStatic
        fun isTeammate(player: Player?): Boolean {
            if (player == null || EntityUtil.isClientPlayer(player)) return false
            val name = player.name.string
            return nameClassMap.containsKey(name)
        }

        @JvmStatic
        fun getColor(dungeonClass: DungeonClass?): Int {
            if (dungeonClass == null) return -0x1

            return when (dungeonClass) {
                ARCHER -> Dungeons.archerColor
                BERSERK -> Dungeons.berserkColor
                HEALER -> Dungeons.healerColor
                MAGE -> Dungeons.mageColor
                TANK -> Dungeons.tankColor
            }
        }

        @JvmStatic
        fun getChar(dungeonClass: DungeonClass?): String {
            if (dungeonClass == null) return "?"

            return when (dungeonClass) {
                ARCHER -> "A"
                BERSERK -> "B"
                HEALER -> "H"
                MAGE -> "M"
                TANK -> "T"
            }
        }

        @JvmStatic
        fun getColor(name: String?): Int = getColor(getClass(name))

        @JvmStatic
        fun isClass(dungeonClass: DungeonClass?): Boolean = currentClass == dungeonClass

        @JvmStatic
        fun isArchTeam(): Boolean = currentClass == ARCHER || currentClass == TANK

        @JvmStatic
        fun isBersTeam(): Boolean = currentClass == BERSERK || currentClass == MAGE

        private fun reset() {
            currentClass = null
            nameClassMap.clear()
        }

        @JvmStatic
        fun printClasses() {
            Misc.addChatMessage(Component.literal("Classes"))
            nameClassMap.forEach { (name, clazz) ->
                Misc.addChatMessage(Component.literal("Name: $name" + "Class: " + clazz.name))
            }
        }
    }
}
